source/_posts/2015-03-26-creating-a-dynamic-single-page-app-for-our-genome-team-using-react.markdown

```diff
@@ -0,0 +1,294 @@
+        <div className="row">
+          ...
+          <div className="col-sm-6">
+            <div className="artist-name">
+              {artistName}
+            </div>
+            <div className="title">
+              {artworkTitle}
+            </div>
+            <div className="date">
+              {artworkDate}
+            </div>
+            ...
+          </div>
+        </div>
+        ...
+      </div>
+    );
+  }
+});
+```
+
+When we implemented a new 'minimized' view for artworks, we also showed the title and artist, and so we broke these bits of information into separate components:
+```
+var ArtistName = React.createClass({
+  handleClick: function() {
+    if (this.props.artwork.artist['name']) {
+      this.props.setTemplateArtistId(this.props.artwork.artist['_id'])
+    }
+  },
+  render: function() {
+    var artistName;
+    if (this.props.artwork.artist && this.props.artwork.artist['name']) {
+      artistName = <a className="artist-as-template-link"
+                      data-template-id={this.props.artwork.artist['_id']}
+                      onClick={this.handleClick}
+                      data-template-type="Artist" href="#">
+                    {this.props.artwork.artist['name']}
+                  </a>;
+    } else if (this.props.artwork['cultural_maker']) {
+      artistName = <span>{this.props.artwork['cultural_maker']}</span>;
+    } else {
+      artistName = <span>Unattributed</span>;
+    }
+    return (
+      <div className="artist-name">
+        {artistName}
+      </div>
+    );
+  }
+});
+
+var ArtworkTitle = React.createClass({
+  render: function() {
+    var artworkTitle = this.props.artwork.title ? this.props.artwork.title : 'Untitled';
+    return (
+      <div className="title">
+        {artworkTitle}
+      </div>
+    );
+  }
+});
+
+var ArtworkDate = React.createClass({
+  render: function() {
+    var artworkDate = this.props.artwork.date ? this.props.artwork.date : 'No Date';
+    return (
+      <div className="date">
+        {artworkDate}
+      </div>
+    );
+  }
+});
+```
+
+And updated our parent to reuse the new child components:
+```
+var PanelArtworkInfo = React.createClass({
+  ...
+  render: function() {
+    ...
+    return (
+      <div className="dummy-wrap">
+        <div className="row">
+          ...
+          <div className="col-sm-6">
+            <ArtistName artwork={this.props.artwork} setTemplateArtistId={this.props.setTemplateArtistId} />
+            <ArtworkTitle artwork={this.props.artwork} />
+            <ArtworkDate artwork={this.props.artwork} />
+            ...
+          </div>
+        </div>
+        ...
+      </div>
+    );
+  }
+});
+```
+
+### Writing Specs
+
+All of the existing specs for the app were written in RSpec, so we chose to write integration tests using RSpec+Capybara. The headless Capybara webkit did not integrate with our React setup, so we switched to Selenium (which also conveniently let us debug our specs within the browser).
```

Comment:
It might not be very obvious what it means to switch from RSpec+Capbara -> Selenium. Personally I have never worked with raw selenium and it sounds scary :)